src/stub_open.c

```diff
@@ -14,21 +15,132 @@
+│     w+      │ O_RDWR | O_CREAT | O_TRUNC    │ FA_CREATE_ALWAYS | FA_WRITE | FA_READ │
+├─────────────┼───────────────────────────────┼───────────────────────────────────────┤
+│     a+      │ O_RDWR | O_CREAT | O_APPEND   | FA_OPEN_APPEND | FA_WRITE | FA_READ   │
+├─────────────┼───────────────────────────────┼───────────────────────────────────────┤
+*/
+static int _map_oflags(int oflag) {
     int result = 0;
-    if (oflag & O_RDONLY) result |= FA_READ;
-    if (oflag & O_WRONLY) result |= FA_WRITE;
-    if (oflag & O_APPEND) result |= FA_OPEN_APPEND;
-    if (oflag & O_CREAT & O_TRUNC) {
-        result |= FA_CREATE_ALWAYS;
-    } else if (oflag & O_CREAT) {
-        result |= FA_CREATE_NEW;
+
+    switch(oflag) {
+        // r
+        case O_RDONLY:
+            result = FA_READ;
+            break;
+        // w
+        case (O_WRONLY | O_CREAT | O_TRUNC):
+            result = FA_CREATE_ALWAYS | FA_WRITE;
+            break;
+        // a
+        case (O_WRONLY | O_CREAT | O_APPEND ):
+            result = FA_OPEN_APPEND | FA_WRITE;
+            break;
+        // r+
+        case (O_RDWR ):
+            result = FA_READ | FA_WRITE;
+            break;
+        // w+
+        case (O_RDWR | O_CREAT | O_TRUNC):
+            result = FA_CREATE_ALWAYS | FA_WRITE | FA_READ;
+            break;
+        // a+
+        case (O_RDWR | O_CREAT | O_APPEND):
+            result = FA_OPEN_APPEND | FA_WRITE | FA_READ;
+            break;
+        default:
+            errno = EINVAL;
+            result = -1;
+            break;
     }
+
     return result;
 }
 
+static void _set_errno(int result) {
+    switch(result) {
+        case FSYS_ERR_DISK_ERR:
+        case FSYS_ERR_INT_ERR:
+        case FSYS_ERR_NOT_READY:
+        case FSYS_ERR_NOT_ENABLED:
+        case FSYS_ERR_NO_FILESYSTEM:
+            errno = ENODEV;
+            break;
+        case FSYS_ERR_NO_FILE:
+        case FSYS_ERR_NO_PATH:
+            errno = ENOENT;
+            break;
+        case FSYS_ERR_INVALID_NAME:
+        case FSYS_ERR_INVALID_OBJECT:
+        case FSYS_ERR_INVALID_DRIVE:
+        case FSYS_ERR_INVALID_PARAMETER:
+            errno = EINVAL;
+            break;
+        case FSYS_ERR_DENIED:
+        case FSYS_ERR_WRITE_PROTECTED:
+            errno = EACCES;
+            break;
+        case FSYS_ERR_EXIST:
+            errno = EEXIST;
+            break;
+        case FSYS_ERR_TOO_MANY_OPEN_FILES:
+            errno = ENFILE;
+            break;
+        case FSYS_ERR_LOCKED:
+            errno = ETXTBSY;
+            break;
+        case FSYS_ERR_MKFS_ABORTED:
+        case FSYS_ERR_TIMEOUT:
+        case FSYS_ERR_NOT_ENOUGH_CORE:
+            errno = EFAULT;
+            break;
+    }
+}
+
 int _Stub_open(const char *path, int oflag, ...) {
-    int m = oflag_to_fatfs(oflag);
-    int result = sys_fsys_open(path, m);
-    return (result >= 0) ? result + FILE_TABLE_OFFSET : result;
+    int mode = _map_oflags(oflag);
+    if(mode < 0) return -1;
```

Comment:
Here it could be considered to use curly braces around the return, but to be honest, I prefer it like it is . Especially since you added a newline after it to separate this logical block from the next. You could add a space after the `if` though, for consistency. There is also a `if` like that below.